### PR TITLE
improve: refactor user data handling, remove rakutenId (SDKCF-3970)

### PR DESCRIPTION
**Pull request by imaginaris**
# Description
This work is based on @apparition47's PR https://github.com/rakutentech/ios-inappmessaging/pull/99

In order to match [Android's API](https://github.com/rakutentech/android-inappmessaging/blob/master/inappmessaging/src/main/java/com/rakuten/tech/mobile/inappmessaging/runtime/UserInfoProvider.kt), `registerPreference()` method has been changed to accept reference to an object that is meant to be updated based on current user session state instead of calling `registerPreference()` with IAMPreference struct each on update.
The SDK will 'read' registered preference object on demand.

RakutenId has been removed to avoid confusion. UserId should be used instead.

## Links
SDKCF-3970

# Checklist
- [X] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [X] I wrote/updated tests for new/changed code
- [X] I removed all sensitive data **before every commit**, including API endpoints and keys
- [X] I ran `fastlane ci` without errors

### CODE REVIEW

File: Tests/UserInfoProviderSpec.swift [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+                beforeEach {
+                    userInfoProviderA = UserInfoProviderMock()
+                    userInfoProviderB = UserInfoProviderMock()
+                }
+
+                it("will return true for nil objects") {
+                    userInfoProviderA = nil
+                    userInfoProviderB = nil
+                    expect(userInfoProviderA == userInfoProviderB).to(beTrue())
```
Comment (apparition47) [inline Tests/UserInfoProviderSpec.swift:66]: Smart. Was overthinking this myself trying to use `equals()`

File: RInAppMessaging/Classes/Enums/Public/Identification.swift [hunk 2]
```diff
@@ -1,6 +1,5 @@
 /// Internal enum for encoding user identifiers
-@objc public enum Identification: Int, Codable {
-    case rakutenId = 1
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Enums/Public/Identification.swift:3]: Maybe we need to check that no customer is relying on this id before removing it.

File: RInAppMessaging/Classes/InAppMessagingModule.swift [hunk 3]
```diff
@@ -164,3 +166,12 @@ extension InAppMessagingModule {
         return randomizer.dice <= config.rolloutPercentage
     }
 }
+
+// MARK: - UserChangeObserver
+extension InAppMessagingModule {
+
+    func didUserChangeOrLogout() {
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/InAppMessagingModule.swift]: didUser -> userDid ?

Is it a checking function or is dealing with a known change of state?

File: RInAppMessaging/Classes/RInAppMessaging.swift [hunk 4]
```diff
@@ -115,16 +115,21 @@ import Foundation
         }
     }
 
-    /// Register user preference to the IAM SDK.
-    /// - Parameter preference: Preferences of the user.
-    @objc public static func registerPreference(_ preference: IAMPreference?) {
+    /// Register user preference object to the IAM SDK.
+    ///
+    /// Registered object should be updated to reflect current user session state.
+    /// There is no need to call this method again, unless new `UserInfoProvider` object has been created.
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/RInAppMessaging.swift]: Maybe this is better:

There is no need to call this method again -> Should only be called once

File: RInAppMessaging/Classes/RInAppMessaging.swift [hunk 5]
```diff
@@ -115,16 +115,21 @@ import Foundation
...
+    /// - Note: This method creates a strong reference to provided object.
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/RInAppMessaging.swift:122]: Nice being explicit about this 👍

**Review by apparition47 — APPROVED** [review]
I see now! Good work!

File: RInAppMessaging/Classes/InAppMessagingModule.swift (hunk 3 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/InAppMessagingModule.swift]: `userDid` is better. It's dealing with a known change

File: RInAppMessaging/Classes/Enums/Public/Identification.swift (hunk 2 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Enums/Public/Identification.swift:3]: Even if they do, they should switch to userId. On backend side userId and rakutenId are treated the same. In the dashboard you provide the list of rakuten ids as user ids. Maybe it's good to mention that in README in migration guide or something like that.

File: RInAppMessaging/Classes/UserInfoProvider.swift [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    case (nil, nil):
+        return true
+    default:
+        return false
+    }
+}
+
+@inlinable
+func != (lhs: UserInfoProvider?, rhs: UserInfoProvider?) -> Bool {
+    !(lhs == rhs)
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/UserInfoProvider.swift:67]: Do you need to unwrap those optionals like in the ==?

File: Tests/UserInfoProviderSpec.swift [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                }
+
+                it("will return false for different userIds") {
+                    userInfoProviderA.userID = "userA"
+                    userInfoProviderB.userID = "userB"
+                    expect(userInfoProviderA == userInfoProviderB).to(beFalse())
+                }
+
+                it("will return true for the same isTrackingIdentifiers") {
```
Comment (donnie-jp) [inline Tests/UserInfoProviderSpec.swift]: isTrackingIdentifiers - idTrackingIdentifiers ?

File: Tests/AccountRepositorySpec.swift [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+                it("will return true if userID was changed") {
+                    let userInfoProvider = UserInfoProviderMock()
+                    userInfoProvider.userID = "userA"
+                    accountRepository.setPreference(userInfoProvider)
+                    accountRepository.updateUserInfo() // first call
+                    userInfoProvider.userID = "userB"
+                    expect(accountRepository.updateUserInfo()).to(beTrue())
+                }
+
+                it("will return false if userID stayed the same changed") {
```
Comment (donnie-jp) [inline Tests/AccountRepositorySpec.swift]: what is "stayed the same changed" meant to be? "stayed the same" ?

File: Tests/AccountRepositorySpec.swift [hunk 9]
```diff
@@ -0,0 +1,283 @@
+import Quick
+import Nimble
+@testable import RInAppMessaging
+
+class AccountRepositorySpec: QuickSpec {
+
+    override func spec() {
+
+        describe("AccountRepository") {
```
Comment (donnie-jp) [inline Tests/AccountRepositorySpec.swift:9]: Very comprehensive unit tests 👍

**Review by donnie-jp — APPROVED** [review]
Looks good 👍

File: RInAppMessaging/Classes/Enums/Public/Identification.swift (hunk 2 above)
Comment (donnie-jp) [reply RInAppMessaging/Classes/Enums/Public/Identification.swift:3]: OK. Yes it'll need a migration guide. The UserInfoProvider change would require a major version anyway so maybe better to include this removal now.

File: RInAppMessaging/Classes/UserInfoProvider.swift (hunk 6 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/UserInfoProvider.swift:67]: `==` implementation will do that for me.

File: Tests/AccountRepositorySpec.swift (hunk 8 above)
Comment (imaginaris) [reply Tests/AccountRepositorySpec.swift]: Haha, yes. Copy&paste leftover 😉

**Pull request merged**